state-and-lifecycle page translated (#60)
reactjs/uz.reactjs.org: content/docs/state-and-lifecycle.md

---
id: state-and-lifecycle
title: Holat va hayotiy ketma-ketlik
permalink: docs/state-and-lifecycle.html
redirect_from:
  - "docs/interactivity-and-dynamic-uis.html"
prev: components-and-props.html
next: handling-events.html
---

> Bu sahifa React komponentidagi holat(state) va hayotiy ketma-ketlik(lifecycle) tushunchalarini tanishtiradi.Bu yerda [batafsil API komponentiga havolani](/docs/react-component.html) topa olasiz.

Reviewer: ```suggestion
Bu sahifa React komponentidagi holat(state) va hayotiy ketma-ketlik(lifecycle) tushunchalarini tanishtiradi.Bu yerda [batafsil komponentinig API ma'lumotnomasini](/docs/react-component.html) topa olasiz.
```

> [Oldingi bo'limlarning bir namunasidagi](/docs/rendering-elements.html#updating-the-rendered-element) soatga e'tibor bering. [Elementlarni chizishda](/docs/rendering-elements.html#rendering-an-element-into-the-dom), Biz faqat foydalanuvchi interfeysini(UI) yangilashni bir yo'lini o'rgandik. Biz chizilgan natijani o'zgartirish uchun `ReactDOM.render()`ni chaqiramiz:

Reviewer: ```suggestion
[Oldingi bo'limlarning namunalardan biridagi](/docs/rendering-elements.html#updating-the-rendered-element) soatga e'tibor bering. [Elementlarni chizish](/docs/rendering-elements.html#rendering-an-element-into-the-dom) bo'limida, biz faqat foydalanuvchi interfeysini(UI) yangilashni bir yo'lini o'rgandik. Biz chizilgan natijani o'zgartirish uchun `ReactDOM.render()`ni chaqiramiz:
```

setInterval(tick, 1000);
```

[**CodePenda sinab ko'ring**](https://codepen.io/gaearon/pen/gwoJZk?editors=0010)

> Bu bo'limda, biz qanday qilib `Clock` komponentini rostdan ham qayta ishlaydigan va jamlangan qilishni o'rganamiz. Soat komponenti hozirgi vaqtni o'rnatadi va har sekundda o'zini yangilaydi.

Reviewer: ```suggestion
Bu bo'limda, biz qanday qilib `Clock` komponentini rostdan ham qayta ishlaydigan va jamlangan qilishni o'rganamiz. Soat komponenti o'zinig taymerini o'rnatadi va har sekundda o'zini yangilaydi.
```

setInterval(tick, 1000);
```

[**CodePenda sinab ko'ring**](https://codepen.io/gaearon/pen/dpdoYR?editors=0010)

> Ammo, u muhim talabni bajarmadi: bu fakt shuki `Clock` taymerni o'rnatadi va foydalanuvchi interfeysini har soniyada yangilaydi, bu `Clock` ning amalga oshirish tafsilotlari bo'lishi kerak.

Reviewer: ```suggestion
Ammo, u muhim talabni bajarmadi: bu fakt shuki taymerni o'rnatish va foydalanuvchi interfeysini har soniyada yangilashini `Clock` komponenti o'zi amalga oshirish kerak.
```

);
```

Buni amalga oshirish uchun, biz `Clock` komponentiga "holat(state)" ni qo'shishimiz kerak.

Holat(State) kiritmalarga(props) o'xshash, lekin u shaxsiy va komponent tomonidan to'liq nazorat qilinadi.

## Funksiyani klassga aylantirish {#converting-a-function-to-a-class}

> Siz `Clock` o'xshash funksional komponentni besh qadamda klass komponentiga aylantira olasiz:

Reviewer: Please use "funksiyaviy komponent" term (see  [TRANSLATION.md](https://github.com/reactjs/uz.reactjs.org/blob/main/TRANSLATION.md#:~:text=function%20component) file)

Reviewer: ```suggestion
Siz `Clock` ga o'xshash funksiyaviy komponentni besh qadamda klass komponentiga aylantira olasiz:
```

1. Bir xil nom bilan `React.Component`dan vorislik oladigan [ES6 klassni](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Classes) yarating.

2. Klassga bitta bo'sh `render()` deb nomlangan metod qo'shing.

3. Funksiyaning tanasini `render()` metodiga ko'chiring.

> 4. `render()` metodi tanasidagi `props` parameterdan olingan o'zgaruvchini `this.props` bilan o'zgartiring.

Reviewer: ```suggestion
4. `render()` metodi tanasidagi `props` o'zgaruvchini `this.props` bilan o'zgartiring.
```

> 5. Qolgan bo'sh funksiya chaqiruvlarini o'chiring.

Reviewer: ```suggestion
5. Qolgan bo'sh funksiyani o'chiring.
```

```js
class Clock extends React.Component {
  render() {
    return (
      <div>
        <h1>Salom, dunyo!</h1>
        <h2>Hozir soat {this.props.date.toLocaleTimeString()}.</h2>
      </div>
    );
  }
}
```

[**CodePenda sinab ko'ring**](https://codepen.io/gaearon/pen/zKRGpo?editors=0010)

`Clock` hozir funksiya sifatida emas klass sifatida belgilangan.

Har doim yangilanish sodir bo'lganida `render` metodi chaqiriladi, lekin agarda biz `<Clock />` komponentini bir DOM tuguniga(node) chizsak, `Clock` klassining faqat bir modeli(instance) foydalaniladi. Bu bizga qo'shimcha imkoniyatlar jumladan lokal holat va hayotiy ketma-ketlik metodlaridan foydalanishga ruhsat beradi.

## Lokal holatni klassga qo'shish {#adding-local-state-to-a-class}

> Biz `date` xususiyatini kiritmalardan(props) holatga(state) uch qadamda ko'chiramiz:

Reviewer: Do we need `xususiyatini` here?

Reviewer: ```suggestion
Biz `date`ni kiritmalardan(props) holatga(state) uch qadamda ko'chiramiz:
```

  render() {
    return (
      <div>
        <h1>Salom, dunyo!</h1>
        <h2>Hozir soat {this.state.date.toLocaleTimeString()}.</h2>
      </div>
    );
  }
}
```

> Qnaday qilib `props`ni asosiy konstruktorga o'tkazganimizni eslab qoling:

Reviewer: ```suggestion
Qnaday qilib `props`ni asosiy konstruktorga o'tkazganimizga etibor bering:
```

);
```

[**CodePenda sinab ko'ring**](https://codepen.io/gaearon/pen/KgQpJd?editors=0010)

Keyin, biz `Clock` komponenti o'zi taymer o'rnatatadigan va har soniyada yangilaydigan qilamiz.

## Adding Lifecycle Methods to a Class {#adding-lifecycle-methods-to-a-class}

> Ko'p komponentlarga ega ilovalarda, komponentlar yo'q qilinganda ular tomonidan olingabn joy va resurslarni

Reviewer: ```suggestion
Ko'p komponentlarga ega ilovalarda, komponentlar yo'q qilinganda ular tomonidan olingan resurslarni
```

> Ko'p komponentlarga ega ilovalarda, komponentlar yo'q qilinganda ular tomonidan olingan resurslarni
> bo'shatish juda muhim.

Reviewer: ```suggestion
Ko'p komponentlarga ega ilovalarda, komponentlar yo'q qilinganda ular tomonidan olingan resurslarni bo'shatish juda muhim.
```

Biz [taymer o'rnatishni](https://developer.mozilla.org/en-US/docs/Web/API/WindowTimers/setInterval) hohlaymiz qachonki `Clock` komponenti DOMga birinchi marta chizilganda va bu Reactda "mounting" deb ataladi.

Shuningdek, biz [taymerni to'xtatishni](https://developer.mozilla.org/en-US/docs/Web/API/WindowTimers/clearInterval) hohlaymiz qachonki `Clock` komponenti tomonidan yaratilgan DOM olib tashlanganda va bu Reactda "unmounting" deb nomlanadi.

> Biz komponent o'rnatilganda(mount) va tozalanganda(unmount) biror kod bajarish uchun klass komponentida maxsus metodlarni e'lon qilishimiz mumkin:

Reviewer: Please use `yechish` for `unmount` [see](https://github.com/reactjs/uz.reactjs.org/blob/main/TRANSLATION.md#:~:text=(un)mount)

Reviewer: ```suggestion
Biz komponent o'rnatilganda(mount) va yechilganda(unmount) biror kod bajarish uchun klass komponentida maxsus metodlarni e'lon qilishimiz mumkin:
```

  render() {
    return (
      <div>
        <h1>Salom, dunyo!</h1>
        <h2>Hozir vaqt {this.state.date.toLocaleTimeString()}.</h2>
      </div>
    );
  }
}
```

Bu metodlar hayotiy ketma-ketlik metodlari("lifecycle methods") deb ataladi.

> `componentDidMount()` metodi komponent natijasi DOMga chizilgandan keyin ishlaydi. Bu taymer o'rnatish uchun juda yacshi joy:

Reviewer: ```suggestion
`componentDidMount()` metodi komponent natijasi DOMga chizilgandan keyin ishlaydi. Bu taymer o'rnatish uchun juda yaxshi joy:
```

  }
```

Taymer IDni qanday saqlaganimizga e'tibor bering (`this.timerID`).

> `this.props` React o'zi tomonidan o'rnatilgan va `this.state` maxsus ma'noga ega paytda, agar siz ma'lumot oqimida qatnashmaydigan narsalarni (masalan, taymer identifikatori) saqlashingiz kerak bo'lsa, siz qo'lda klassga qo'shimcha maydonlarni qo'shishingiz mumkin..

Reviewer: ```suggestion
`this.props` React o'zi tomonidan o'rnatilgan va `this.state` maxsus ma'noga ega. Agar siz ma'lumot oqimida qatnashmaydigan narsalarni (masalan, taymer identifikatori) saqlashingiz kerak bo'lsa, siz qo'lda klassga qo'shimcha maydonlarni qo'shishingiz mumkin..
```

Taymerni `componentWillUnmount()` hayotiy ketma-ketlik metodida yo'q qilamiz:

```js{2}
  componentWillUnmount() {
    clearInterval(this.timerID);
  }
```

Va nihoyat, biz `Clock` komponentini har soniyada ishlatadigan `tick()` metodini qo'llaymiz.

> Komponent lokal holatiga yangilanishlarni rejalashtirish uchun u `this.setState ()` dan foydalanadi:

Reviewer: ```suggestion
Komponent lokal holatiga yangilanishlarni rejalashtirish uchun u `this.setState()` dan foydalanadi:
```

);
```

[**CodePenda sinab ko'ring**](https://codepen.io/gaearon/pen/amqdNA?editors=0010)

Endi soat har soniyada ishlaydi.

Keling, nima bo'layotganini va metodlar chaqirilish tartibini tezlik bilan takrorlaylik:

> 1) `<Clock />` komponenti `ReactDOM.render ()` ga o'tkazilganda, React `Clock` komponentining konstruktorini chaqiradi. `Clock` joriy vaqtni ko'rsatishi kerak bo'lganligi sababli, u "this.state" ni joriy vaqtni o'z ichiga olgan ob'ekt bilan boshlaydi. Keyinchalik bu holatni yangilaymiz.

Reviewer: ```suggestion
1) `<Clock />` komponenti `ReactDOM.render()` ga o'tkazilganda, React `Clock` komponentining konstruktorini chaqiradi. `Clock` joriy vaqtni ko'rsatishi kerak bo'lganligi sababli, u "this.state" ni joriy vaqtni o'z ichiga olgan ob'ekt bilan boshlaydi. Keyinchalik bu holatni yangilaymiz.
```

2) React keyin `Clock` komponentining `render()` metodini chaqiradi. Shunday qilib, React ekranda nimani ko'rsatish kerakligini bilib oladi. React keyin DOMni `Clock` komponenti ko'rsatuviga mos keladigan tarzda yangilaydi.

3) DOMga `Clock` natijasi kiritilganda, React `componentDidMount()` hayotiy ketma-ketlik metodini chaqiradi. Uning ichida `Clock` komponenti brauzerdan sekundiga bir marta `tick()` metodini chaqirish uchun taymerni o'rnatishni so'raydi.

> 4) Har soniyada brauzer `tick()` metodini chaqiradi. Uning ichida `Clock` komponenti joriy vaqtni o'z ichiga olgan ob'ekt bilan `setState()`ni chaqirib, UI yangilanishini rejalashtiradi. `SetState()` chaqiruvi tufayli React holat o'zgarganini biladi va ekranda nima bo'lishi kerakligini bilish uchun `render()` usulini qayta chaqiradi. Bu safar `render()` usulidagi `this.state.date` har xil bo'ladi, shuning uchun render chiqishi yangilangan vaqtni o'z ichiga oladi. React DOMni shunga mos ravishda yangilaydi.

Reviewer: ```suggestion
4) Har soniyada brauzer `tick()` metodini chaqiradi. Uning ichida `Clock` komponenti joriy vaqtni o'z ichiga olgan ob'ekt bilan `setState()`ni chaqirib, UI yangilanishini rejalashtiradi. `setState()` chaqiruvi tufayli React holat o'zgarganini biladi va ekranda nima bo'lishi kerakligini bilish uchun `render()` usulini qayta chaqiradi. Bu safar `render()` usulidagi `this.state.date` har xil bo'ladi, shuning uchun renderning qaytargan natijasi yangilangan vaqtni o'z ichiga oladi. React DOMni shunga mos ravishda yangilaydi.
```

);
```

[**CodePenda sinab ko'ring**](https://codepen.io/gaearon/pen/vXdGmd?editors=0010)

Har bir `Clock` o'z taymerini o'rnatadi va mustaqil ravishda yangilanadi.

> React ilovalarida komponentning holatli yoki holatsiz ekanligi vaqt o'tishi bilan o'zgarishi mumkin bo'lgan komponentni amalga oshirish tafsiloti hisoblanadi. Holatsiz bo'lgan komponentlarni holatli komponentlar ichida ishlatishingiz mumkin va aksincha.

Reviewer: ```suggestion
React ilovalarida komponentning holatli yoki holatsiz ekanligi komponentni amalga oshirish tafsiloti hisoblanadi va bu vaqt o'tishi bilan o'zgarishi mumkin. Holatsiz bo'lgan komponentlarni holatli komponentlar ichida ishlatishingiz mumkin va aksincha.
```
